### ISSUE TITLE
Unstable App with Limited Features

### ISSUE THREAD

**Issue by thiagojramos** [civitai/civitai-link-desktop#22]
Every interaction I have with the app results in it automatically closing itself, disappearing into the systray, as well as not showing any label on the taskbar when the window is in focus, just an icon in the tray. When minimized or out of focus, instead of staying active in the background with the window open, it completely hides itself.

The app's performance is lacking, operating with notable slowness (Print 01). Its functionality is limited to a list of models located in the specified folders, without offering any form of categorization or additional detail of the models, which is in stark contrast with the intuitive and information-rich navigation available in the Stability Matrix. Additionally, the app window is not resizable, which is a nuisance for someone with a large screen (21:9), as the interface appears too small.

<details><summary>Print 01</summary>
<p>

![Print 01](https://github.com/civitai/civitai-link-desktop/assets/45890502/4c55fd44-d9ea-49c9-bb63-f0ff95bc25b3)

</p>
</details> 

Another limitation is the absence of a preview image download and the creation of a .json file containing information about the models (Print 02 and 03). This feature is particularly valuable for WebUI users (like automatic, forge, etc), who utilize these data in their interfaces.

<details><summary>Print 02</summary>
<p>

![Print 02](https://github.com/civitai/civitai-link-desktop/assets/45890502/a4b1d57b-995e-4387-99cd-47130bb997ad)

</p>
</details> 
<details><summary>Print 03</summary>
<p>

![Print 03](https://github.com/civitai/civitai-link-desktop/assets/45890502/49de78c3-bbfd-462a-bf6e-3d57410c2014)

</p>
</details> 

Some models that I have previously downloaded are listed on the site as if I had not downloaded them (Print 04), and when I click on the button, it shows an error and says that it has already been downloaded (Print 05).

<details><summary>Print 04</summary>
<p>

![Print 04](https://github.com/civitai/civitai-link-desktop/assets/45890502/088f429e-ff4a-4d7b-b1bc-333023701308)

</p>
</details> 

<details><summary>Print 05</summary>
<p>

![Print 05](https://github.com/civitai/civitai-link-desktop/assets/45890502/9a5d9015-5578-40a4-bbec-7c8f5e246edc)

</p>
</details> 

This feedback is intended not only to point out areas for improvement but also to reinforce the importance of a refined and efficient user experience, which is currently hindered by these issues.

**Comment by thiagojramos** [comment]
Regrettably, after updating to version "1.13.4" for Windows, things seem to have worsened rather than improved. Despite the hope for fixes to previously reported issues, I now face new challenges in addition to the already known ones.

During an attempt to download a model, the application started downloading the same file twice simultaneously, both with exasperating slowness. This is particularly frustrating when compared to the "less worse" Civitai Link extension for WebUI, which performs downloads at the appropriate speed. The downloaded files, to make matters worse, are incomplete, not reaching even half the expected size as indicated on the site; one of them had only a few kilobytes. Additionally, both remained in the "temp" folder, without being moved to the "lora" folder upon download completion.

One of the supposed novelties of this version would be the generation of a .json file containing data about the downloaded model. However, the provided data is, at best, incomplete—or even useless. The file failed to identify the version of the Stable Diffusion of the model, listing it as "unknown", and omitted critical information such as the "activation text". Essentially, the generated .json was reduced to this:

`{"sd version":"unknown","preferred weight":0.8,"notes":""}`

Note: as you can see in the print below, the model was an SDXL, hence recognizable. I say this because the extension for the WebUI only failed to recognize pony models.

There is still no functionality for downloading preview images, and the other bugs and utilities previously reported remain unresolved. The disappointment with the development and improvement of the application is palpable, reflecting a user experience that leaves much to be desired.

<details><summary>Prints</summary>
<p>

![2024-04-09 (4)](https://github.com/civitai/civitai-link-desktop/assets/45890502/ba4e8b2f-7bf1-4e4c-b14b-8d867f095ae3)
![2024-04-09 (5)](https://github.com/civitai/civitai-link-desktop/assets/45890502/ea34fb45-a105-468d-b38f-3e0c828d43e5)
![2024-04-09 (3)](https://github.com/civitai/civitai-link-desktop/assets/45890502/66840b88-ddaa-4e54-be50-aeadf01c9fcc)

</p>
</details>

**Comment by cmoody** [comment]
Thanks for the feedback @thiagojramos. A new UI is coming out and I am hoping to address a bunch of these issues. Please be patient as this takes time to both replicate, debug and fix. Im hoping to get this updated and out this week.

I will synthesize this as a punch list and update here as I get them updated. The biggest focus at the moment is around the performance issues with large amounts of files. I have some ideas on whats happening with a change made to try to improve how we check and hash files.

**Comment by cmoody** [comment]
Hey @thiagojramos this is kind of the list of what I noted from your comments above. Let me know if I didnt capture something from it. 1.14.0 just went out and hopefully addresses a lot of the problems you were experiencing. Theres still a lot more that Im looking into for performance and will be updating but Im hoping this unblocks everyone.

Open Issues
- Performance slowness (investigating)
- Specified folders and categorization (Can you clarify a bit?)
- Already downloaded w/ error (Possibly fixed with new UI and fire event)

New UI (1.14.0)
- System tray app, minimize, focus, icon (Fixed in new UI)
- Additional model details (Fixed in new UI)
- Preview image and json file for A1111 (Fixed in new UI)
- Resizable window (Fixed in new UI)
- Fire event type: 'resources:list' when done scanning file system (Fixed in new UI)

**Comment by thiagojramos** [comment]
Hey, @cmoody thanks for the update and for addressing the issues so promptly! I really appreciate the effort to enhance the app's functionality and performance.

Regarding the "Specified folders and categorization", it would be great to have some sort of filter to view only specific types of models, such as LoRa, Checkpoints, etc. This would help in navigating through large collections of models more efficiently. If possible, a block-style display mode and a right-click context menu with options like "View in Explorer" would be incredibly useful (if these aren't already included in the new version).

However, I'm experiencing a new issue with the latest update (1.14.2). After installing it and entering the link key, the screen turns completely white. I've tried reinstalling multiple times, rebooting my PC, and clearing all entries, but nothing seems to fix it. There was an option for "Dev Tools" in the systray, and I've attached a print screen below showing the error in the console tab.

![2024-04-11](https://github.com/civitai/civitai-link-desktop/assets/45890502/be0c7b5c-4e2c-4cd2-91e9-9f9bdb132978)


I hope this can be looked into as well. Thanks again for your responsiveness and looking forward to the continuous improvements.

**Comment by cmoody** [comment]
Thanks for the screenshot of the devtools. Im glad I put that in there! I think I know whats happening and hopefully have that fixed up in the latest patch.

Ive got an idea around some sorting/filtering for different categorizations but I like where you are thinking on that. I talk with design and see how we can get that to be a bit more useful.

**Comment by thiagojramos** [comment]
@cmoody, Thanks so much for the quick fix and for considering my feedback on the sorting/filtering options! I'm really pleased to hear you're exploring those ideas.

I'm happy to report that the installation of version 1.14.3 went smoothly and the app is now functioning properly. The new features are a significant improvement and are working well.

However, I've noticed that the app is still using a bit more memory than expected, around 320-350MB, even when minimized to the systray (not actively doing anything). Just thought I'd mention it as you continue to look into performance enhancements.

Thanks again for your responsiveness and for the continuous updates. It's great to see such proactive support!

**Comment by cmoody** [comment]
@thiagojramos good to know. Im hoping to start focusing on the memory usage issues after an update going out today or tomorrow. My guess is that with the way the data store is working that it might be holding it all in memory. Im hoping to explore some other options that will make loading data more on demand. Will keep you udpated.

**Comment by thiagojramos** [comment]
Thanks for the update, @cmoody! I'm glad to hear that the memory usage issues are on your radar for future optimizations. I appreciate your efforts in continuously improving the app.

I've noticed something concerning the .json files created for the models (especially LoRa) by the Link App: they do not include the "activation text" that the SD UI refers to as "Trigger Words" on the website. I'm attaching two screenshots illustrating these points: one shows a .json file without the "activation text"/"Trigger Words" downloaded through the Link App, and the other with them (and the file hash that SD identified it) downloaded via the old Civitai extension for SD.

<details><summary>LoRa without Activation Text (LinkApp)</summary>
<p>

![LoRa-Without-Activation-Text](https://github.com/civitai/civitai-link-desktop/assets/45890502/9006ee40-e75e-4394-90df-7e07c751a9fc)

</p>
</details> 

<details><summary>LoRa with Activation Text (Old SD ext.)</summary>
<p>

![LoRa-With-Activation-Text](https://github.com/civitai/civitai-link-desktop/assets/45890502/8687edd8-334a-48de-b228-4b871b4b21dc)

</p>
</details> 

Note: Just for clarification, the text might appear in Portuguese in my screenshots due to a userscript I use to translate the Civitai interface, but "Texto de ativação" in the SD UI and "Palavras de Gatilho" on the site are known as "activation text" and "Trigger Words" in English, respectively.


Less importantly, I've noticed that when I search for the name of a LoRa model in the Link App that I had downloaded using the SD extension, even though it's in the same folder as the models downloaded through the Link App, it only appears after a significant delay and after I've opened and closed the Link App. A potential improvement might be to add a "refresh list" button somewhere at the top of the window.

Also, when the program is in the foreground, it would be beneficial for the icon/label of the program to appear in the taskbar, mainly for quick access to Windows shortcuts like right-click to close the app more swiftly, pin the app to the taskbar, and end task to avoid opening the task manager to force the program to close. I'm not sure if it's related, but when I use Alt+Tab to switch windows, the Link App does not appear among the open programs (when it's not hidden in the systray).

Thanks again for your attention and for keeping the community informed about ongoing improvements!

**Comment by cmoody** [comment]
@thiagojramos I think in 1.15.0 the activation text should be in the json file. I think I was missing passing it down to the function that creates it.

> Less importantly, I've noticed that when I search for the name of a LoRa model in the Link App that I had downloaded using the SD extension, even though it's in the same folder as the models downloaded through the Link App, it only appears after a significant delay and after I've opened and closed the Link App. A potential improvement might be to add a "refresh list" button somewhere at the top of the window.

Ill take a look and see if we can improve that. Ive got a few refresh button ideas for various things including Vault that will be coming pretty soon.

> Also, when the program is in the foreground, it would be beneficial for the icon/label of the program to appear in the taskbar, mainly for quick access to Windows shortcuts like right-click to close the app more swiftly, pin the app to the taskbar, and end task to avoid opening the task manager to force the program to close. I'm not sure if it's related, but when I use Alt+Tab to switch windows, the Link App does not appear among the open programs (when it's not hidden in the systray).

I think this should be working now if you can verify for me.

**Comment by cmoody** [comment]
> In some models, the preview doesn't download. 
I think this might be due to it being animated. Im trying to get some resolution around this but its been a bit slow going.

> Sorting issue
Ill take a look. Ive noticed its a little buggy at times. I think I need to rework how the data refreshes.

> It might be helpful to have a progress bar somewhere, possibly even as a small popup in the corner of the app screen, when the program is scanning installed models.
I think it freezes because technically progress is blocking the main thread. Ive been looking at worker threads to help offload that work. Theres just some questions I have around how it reports back to Electron so still investigating.

> Download stuck
Theres an issue that sometimes happens with images that is tracked in another github issue. Im going to test out a new download package to see if that helps and if so Im going to switch up the file download as well. 🤞 that works

> Returning to the .json files, for Pony models, it continues to list "unknown" for the type.
I took the code from the SD extension 🤣  Its just an if statement that only checks for 3 things. Should be an easy fix.

Thanks for all the feedback. Its definitely helping make big strides in getting this more stable and into a better spot for everyone.

**Comment by thiagojramos** [comment]
About the previews that won't download. Initially, I thought it might be because of the image format on the website (.jpeg). Because I noticed that all the models that downloaded with a preview were .png, but when opening the image of [Dr. Eggman's LoRa](https://civitai.com/models/422613/dr-eggman-pony-xl), IrfanView immediately warned that it was incorrect.

<details><summary>Eggman</summary>
<p>

![2024-04-29 (4)](https://github.com/civitai/civitai-link-desktop/assets/45890502/17eaff81-7286-461c-9d3b-483e7eaa4fd1)

</p>
</details> 

And it's not just about only gif/mp4 or nsfw previews on the site because the one of the "[Pirate that stretches and dies if he takes a bath](https://civitai.com/models/423158?modelVersionId=471466)", the image is static and the site doesn't even tag the image as nsfw.

<details><summary>Pirate boy</summary>
<p>

![2024-04-29 (1)](https://github.com/civitai/civitai-link-desktop/assets/45890502/3b8d03bc-2169-423c-a0d6-d1af3d2bff6b)

</p>
</details> 

Regarding the animated ones, the Link App is downloading the next image (which is not animated) normally. Here's an example with the LoRa from [Alice In Wonderland](https://civitai.com/models/35930/alice-in-wonderland-disney-sd-15-or-xl-pony-by-yeiyeiart), where the author mostly uses the first animated image in their LoRas.

<details><summary>Alice</summary>
<p>

![2024-04-29 (6)](https://github.com/civitai/civitai-link-desktop/assets/45890502/5176585c-d449-442e-8b8b-479075b32d3d)

</p>
</details>

**Comment by cmoody** [comment]
@thiagojramos sorry for the delay. Been heads down on some fixes. Great catch! I was hard coding something and it was labeling things with the wrong extension. Ive got that fixed and going out in the next update.

**Comment by cmoody** [comment]
Hey @thiagojramos ! Thanks for your patience on this. We just did a big overhaul to how we hash files and watch folders for models. Hoping this helps alleviate a lot of the slow startup issues and getting stuck on initial setup. Your app should automatically update to 1.18.0 but it might be worth reinstalling just so it doesnt try to start the download process in 1.16.1.
